File: demos/realtime_motion_graph/client/protocol.py
"""WebSocket protocol shared by the client, the server, and the full demo.

Torch-free. Only depends on numpy, websockets, and zstandard.
"""

import json
import struct
import threading

import numpy as np

SAMPLE_RATE = 48000
T = 1500  # 60s at 25fps latents
CROSSFADE_SECONDS = 0.05

# Binary slice header: flags, start_sample, num_samples, channels, tick_ms, dec_ms, num_gens
# flags: 0 = raw float16, 1 = zstd-compressed float16 delta
SLICE_HDR_FMT = "<BIIHffI"
SLICE_HDR_SIZE = struct.calcsize(SLICE_HDR_FMT)
SLICE_FLAG_RAW = 0
SLICE_FLAG_DELTA = 1


class RemoteBackend:
    """WebSocket connection to a remote GPU pipeline server.

    Accepts the source audio as either a torch.Tensor with shape
    ``(channels, samples)`` (via ``.numpy()``) or a plain numpy array with
    the same shape. This keeps the thin client torch-free while remaining
    compatible with the full local+remote demo.
    """

    def __init__(self, url: str, waveform, config: dict):
        from websockets.sync.client import connect as ws_connect

        print(f"[Remote] Connecting to {url}...")
        self.ws = ws_connect(url, max_size=50 * 1024 * 1024, open_timeout=30)
        self._send_lock = threading.Lock()

        print("[Remote] Sending config...")
        self.ws.send(json.dumps(config))

        if hasattr(waveform, "numpy"):
            audio_np = waveform.numpy()
        else:
            audio_np = np.asarray(waveform)
        # Normalize to (samples, channels) layout for the wire format.
        # Callers pass (channels, samples); transpose here.
        audio_np = audio_np.T
        hdr = struct.pack("<II", audio_np.shape[1], audio_np.shape[0])
        self.ws.send(hdr + audio_np.astype(np.float32).tobytes())
        print(f"[Remote] Uploaded audio ({audio_np.nbytes / 1024 / 1024:.1f} MB)")

        print("[Remote] Waiting for server init...")
        ready = json.loads(self.ws.recv())
        self.duration = ready["duration"]
        self.channels = ready["channels"]
        self.sample_rate = ready["sample_rate"]
        self.lora_count = ready.get("lora_count", 0)

        init_bytes = self.ws.recv()
        self.initial_buffer = (
            np.frombuffer(init_bytes, dtype=np.float16)
            .astype(np.float32)
            .reshape(-1, self.channels)
        )
        print(f"[Remote] Ready: {self.duration:.1f}s, {self.channels}ch")

    def send_raw(self, raw_dict: dict, playback_pos: float):
        msg = {"type": "params", "raw": raw_dict, "playback_pos": playback_pos}
        try:
            with self._send_lock:
                self.ws.send(json.dumps(msg))
        except Exception:
            pass

    def send_prompt(self, tags: str):
        try:
            with self._send_lock:
                self.ws.send(json.dumps({"type": "prompt", "tags": tags}))
        except Exception:
            pass

    def recv(self, timeout: float = 0.01):
        try:
            msg = self.ws.recv(timeout=timeout)
            if isinstance(msg, bytes) and len(msg) > SLICE_HDR_SIZE:
                hdr = struct.unpack(SLICE_HDR_FMT, msg[:SLICE_HDR_SIZE])
                flags = hdr[0]
                payload = msg[SLICE_HDR_SIZE:]
                if flags == SLICE_FLAG_DELTA:
                    import zstandard as zstd
                    payload = zstd.decompress(payload)
                audio_f16 = np.frombuffer(payload, dtype=np.float16)
                return ("audio", {
                    "flags": flags,
                    "start_sample": hdr[1],
                    "num_samples": hdr[2],
                    "channels": hdr[3],
                    "tick_ms": hdr[4],
                    "dec_ms": hdr[5],
                    "num_gens": hdr[6],
                    "audio": audio_f16.astype(np.float32).reshape(hdr[2], hdr[3]),
                })
            elif isinstance(msg, str):
                return ("json", json.loads(msg))
        except TimeoutError:
            return None
        except Exception:
            return None
        return None

    def close(self):
        try:
            self.ws.close()
        except Exception:
            pass

File: demos/realtime_motion_graph_web/static/protocol.js
// WebSocket wire protocol for the realtime motion-to-music demo.
//
// Mirrors demos/realtime_motion_graph/client/protocol.py:
//   - init: json config, then binary (uint32 channels, uint32 samples) + float32 PCM
//   - server sends: json ready, then binary initial buffer (float16)
//   - streaming: json params/prompt out, binary slices + json params_update in
//
// Binary slice header:  '<BIIHffI'  (little-endian)
//   uint8  flags (0 raw, 1 zstd-delta)
//   uint32 start_sample
//   uint32 num_samples
//   uint16 channels
//   float32 tick_ms
//   float32 dec_ms
//   uint32 num_gens

export const SAMPLE_RATE = 48000;
export const T = 1500;                         // 60s at 25fps latents
export const CROSSFADE_SECONDS = 0.05;

export const SLICE_HDR_SIZE = 1 + 4 + 4 + 2 + 4 + 4 + 4;  // = 23
export const SLICE_FLAG_RAW = 0;
export const SLICE_FLAG_DELTA = 1;

// --- float16 -> float32 ------------------------------------------------
// Browsers don't have native float16, so decode by hand via a reusable
// Uint32Array/Float32Array overlay to avoid per-sample object churn.

const _fBuf = new ArrayBuffer(4);
const _fU32 = new Uint32Array(_fBuf);
const _fF32 = new Float32Array(_fBuf);

function _half2single(h) {
  const s = (h & 0x8000) << 16;
  let e = (h & 0x7c00) >> 10;
  let f = h & 0x03ff;
  if (e === 0) {
    if (f === 0) { _fU32[0] = s; return _fF32[0]; }
    // subnormal -> normalise
    while ((f & 0x0400) === 0) { f <<= 1; e--; }
    e++;
    f &= ~0x0400;
  } else if (e === 31) {
    _fU32[0] = s | 0x7f800000 | (f << 13);
    return _fF32[0];
  }
  e = e + (127 - 15);
  _fU32[0] = s | (e << 23) | (f << 13);
  return _fF32[0];
}

export function float16ArrayToFloat32(u16) {
  const out = new Float32Array(u16.length);
  for (let i = 0; i < u16.length; i++) out[i] = _half2single(u16[i]);
  return out;
}

// --- zstd decompression ------------------------------------------------
// fzstd is loaded as a global via <script> in index.html.  Its UMD build
// exposes ``fzstd.decompress(Uint8Array) -> Uint8Array``.

function _decompressZstd(bytes) {
  if (typeof fzstd === "undefined" || !fzstd.decompress) {
    throw new Error("fzstd library not loaded - check lib/fzstd.min.js or network");
  }
  return fzstd.decompress(bytes);
}

// --- RemoteBackend -----------------------------------------------------

export class RemoteBackend extends EventTarget {
  /**
   * @param {string} url
   * @param {Float32Array} interleaved  (samples * channels,) float32
   * @param {number} channels
   * @param {object} config
   */
  constructor(url, interleaved, channels, config) {
    super();
    this.url = url;
    this._pending = { interleaved, channels, config };
    this.ready = false;
    this.initialBuffer = null;  // Float32Array interleaved
    this.duration = 0;
    this.channels = 0;
    this.sampleRate = SAMPLE_RATE;
  }

  async connect() {
    return new Promise((resolve, reject) => {
      const ws = new WebSocket(this.url);
      ws.binaryType = "arraybuffer";
      this.ws = ws;

      let phase = "config";  // "config" -> "init-binary" -> "streaming"

      ws.onopen = () => {
        // Phase 1: send JSON config and then binary audio upload.
        ws.send(JSON.stringify(this._pending.config));
        const { interleaved, channels } = this._pending;
        const samples = interleaved.length / channels;
        const hdr = new ArrayBuffer(8);
        const dv = new DataView(hdr);
        dv.setUint32(0, channels, true);
        dv.setUint32(4, samples, true);
        // Concatenate header + float32 PCM into one ArrayBuffer.
        const pcm = new Uint8Array(interleaved.buffer);
        const combined = new Uint8Array(hdr.byteLength + pcm.byteLength);
        combined.set(new Uint8Array(hdr), 0);
        combined.set(pcm, hdr.byteLength);
        ws.send(combined);
        phase = "ready";
      };

      ws.onmessage = (ev) => {
        if (phase === "ready") {
          // Expect JSON {"type":"ready",...}
          try {
            const msg = JSON.parse(ev.data);
            if (msg.type !== "ready") {
              reject(new Error(`Unexpected init message: ${ev.data}`));
              return;
            }
            this.duration = msg.duration;
            this.channels = msg.channels;
            this.sampleRate = msg.sample_rate;
            phase = "initial-buffer";
          } catch (e) { reject(e); }
          return;
        }

        if (phase === "initial-buffer") {
          // Binary float16 initial buffer.
          const u16 = new Uint16Array(ev.data);
          this.initialBuffer = float16ArrayToFloat32(u16);
          this.ready = true;
          phase = "streaming";
          this._pending = null;
          resolve(this);
          this.dispatchEvent(new CustomEvent("ready"));
          return;
        }

        // --- Streaming phase ---
        if (typeof ev.data === "string") {
          let msg;
          try { msg = JSON.parse(ev.data); } catch { return; }
          if (msg.type === "params_update") {
            this.dispatchEvent(new CustomEvent("params", { detail: msg.params }));
          } else if (msg.type === "prompt_applied") {
            this.dispatchEvent(new CustomEvent("prompt_applied", { detail: msg.tags }));
          } else {
            this.dispatchEvent(new CustomEvent("json", { detail: msg }));
          }
          return;
        }

        // Binary slice.
        try {
          const slice = this._parseSlice(ev.data);
          if (slice) this.dispatchEvent(new CustomEvent("slice", { detail: slice }));
        } catch (e) {
          console.error("[protocol] slice parse failed:", e);
        }
      };

      ws.onerror = (e) => {
        console.error("[protocol] ws error", e);
        if (!this.ready) reject(new Error("WebSocket connection failed (network / port unreachable)"));
        this.dispatchEvent(new CustomEvent("error", { detail: e }));
      };
      ws.onclose = (e) => {
        // If the socket closes before we finished the init handshake,
        // the connect() promise must reject — otherwise the launcher
        // sits on "Uploading..." forever when the server crashes mid-init.
        if (!this.ready) {
          const reason = e.reason || `code ${e.code}`;
          reject(new Error(`WebSocket closed before server sent 'ready' (${reason}). Check the server console.`));
        }
        this.dispatchEvent(new CustomEvent("close", { detail: e }));
      };
    });
  }

  _parseSlice(buf) {
    if (buf.byteLength < SLICE_HDR_SIZE) return null;
    const dv = new DataView(buf);
    let o = 0;
    const flags       = dv.getUint8(o); o += 1;
    const startSample = dv.getUint32(o, true); o += 4;
    const numSamples  = dv.getUint32(o, true); o += 4;
    const channels    = dv.getUint16(o, true); o += 2;
    const tickMs      = dv.getFloat32(o, true); o += 4;
    const decMs       = dv.getFloat32(o, true); o += 4;
    const numGens     = dv.getUint32(o, true); o += 4;

    let payload = new Uint8Array(buf, SLICE_HDR_SIZE);
    if (flags === SLICE_FLAG_DELTA) {
      payload = _decompressZstd(payload);
    }
    // Copy so the Uint16Array is 2-byte aligned regardless of the
    // underlying buffer's origin (zstd output has its own backing).
    const aligned = new ArrayBuffer(payload.byteLength);
    new Uint8Array(aligned).set(payload);
    const u16 = new Uint16Array(aligned);
    const audio = float16ArrayToFloat32(u16);

    return {
      flags, startSample, numSamples, channels,
      tickMs, decMs, numGens, audio,
    };
  }

  sendParams(raw, playbackPos) {
    if (this.ws?.readyState !== WebSocket.OPEN) return;
    try {
      this.ws.send(JSON.stringify({
        type: "params",
        raw,
        playback_pos: playbackPos,
      }));
    } catch {}
  }

  sendPrompt(tags, key) {
    if (this.ws?.readyState !== WebSocket.OPEN) return;
    try {
      const msg = { type: "prompt", tags };
      if (key) msg.key = key;
      this.ws.send(JSON.stringify(msg));
    } catch {}
  }

  close() {
    try { this.ws?.close(); } catch {}
  }
}
